crates/sui-core/src/authority/authority_per_epoch_store.rs

```diff
@@ -2538,29 +2571,51 @@ impl AuthorityPerEpochStore {
             Some(rm) => Some(rm.lock().await),
             None => None,
         };
-        let dkg_closed = self.randomness_state_enabled()
+        let dkg_failed = self.randomness_state_enabled()
             && randomness_manager
                 .as_ref()
                 .expect("randomness manager should exist if randomness is enabled")
-                .is_dkg_closed();
-        if dkg_closed {
-            let deferred_randomness_txs =
-                self.load_deferred_transactions_for_randomness(&mut batch)?;
-            previously_deferred_tx_digests.extend(deferred_randomness_txs.iter().flat_map(
-                |(deferral_key, txs)| {
-                    txs.iter().map(|tx| match tx.0.transaction.key() {
-                        SequencedConsensusTransactionKey::External(
-                            ConsensusTransactionKey::Certificate(digest),
-                        ) => (digest, *deferral_key),
-                        _ => panic!(
-                            "deferred randomness transaction was not a user certificate: {tx:?}"
-                        ),
-                    })
-                },
-            ));
-            sequenced_randomness_transactions
-                .extend(deferred_randomness_txs.into_iter().flat_map(|(_, txs)| txs));
+                .is_dkg_failed();
+        if dkg_failed {
+            self.load_and_process_deferred_transactions_for_randomness(
+                &mut batch,
+                &mut previously_deferred_tx_digests,
+                &mut sequenced_randomness_transactions,
+            )?;
         }
+
+        // Generate randomness for this commit if:
+        // 1. randomness is enabled
+        // 2. DKG is completed successfully
+        // 3. we are still accepting certs
+        let randomness_round = if !dkg_failed
+            && self.randomness_state_enabled()
+            && randomness_manager
+                .as_ref()
+                .expect("randomness manager should exist if randomness is enabled")
+                .is_dkg_successful()
```

Comment:
this logic, (from line 2574 down) is getting pretty confusing. I think I understand it, but I think it might be worth making the randomness state machine a bit more explicit, i.e. with an enum representing the different states (DkgInProgress, DkgFailed, DkgSucceeded, etc). Then, perhaps we can have one big match where the process for each state is carried out without a ton of conditionals interspersed.